Issues with gU on qute://help/

Doing `gU` on `qute://help/quickstart.html` tries to open `qute://help/quickstart.html/..`.

In https://crashes.qutebrowser.org/view/a07c404e this also lead to an `AssertionError` on 984dd1ba8:

```
commands   command:run:489 command called: navigate ['up', '-t']
commands   command:run:504 Calling qutebrowser.browser.commands.CommandDispatcher.navigate(<qutebrowser.browser.commands.CommandDispatcher>, 'up', True, False, False, 1)
webview    tabbedbrowser:tabopen:445 Creating new tab with URL PyQt5.QtCore.QUrl('qute://help/img/cheatsheet-big.png/..'), background False, related True, idx None
[...]
webview    webview:acceptNavigationRequest:296 navigation request: url qute://help/img/cheatsheet-big.png/.., type NavigationTypeTyped, is_main_frame True
misc       webenginequtescheme:requestStarted:50 Got request for qute://help/img/cheatsheet-big.png/..
misc       qutescheme:data_for_url:143 url: qute://help/img/cheatsheet-big.png/.., path: /img/cheatsheet-big.png/.., host help
misc       crashsignal:exception_hook:216 Uncaught exception
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/webengine/webenginequtescheme.py", line 52, in requestStarted
    mimetype, data = qutescheme.data_for_url(url)
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/qutescheme.py", line 158, in data_for_url
    mimetype, data = handler(url)
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/qutescheme.py", line 118, in wrapper
    return self._function(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/qutescheme.py", line 329, in qute_help
    assert mimetype is not None, url
AssertionError: PyQt5.QtCore.QUrl('qute://help/img/cheatsheet-big.png/..')
```

--- a/qutebrowser/browser/navigate.py
+++ b/qutebrowser/browser/navigate.py
@@ -64,6 +64,7 @@ def path_up(url, count):
         raise Error("Can't go up!")
     for _i in range(0, min(count, path.count('/'))):
         path = posixpath.join(path, posixpath.pardir)
+    path = posixpath.normpath(path)
     url.setPath(path)
     return url
 

--- a/qutebrowser/browser/qutescheme.py
+++ b/qutebrowser/browser/qutescheme.py
@@ -135,6 +135,11 @@ def data_for_url(url):
     Return:
         A (mimetype, data) tuple.
     """
+    # normalize path and strip redundant trailing slashes
+    norm_url = url.adjusted(QUrl.NormalizePathSegments | QUrl.StripTrailingSlash)
+    if norm_url != url:
+       raise Redirect(norm_url)
+
     path = url.path()
     host = url.host()
     query = urlutils.query_string(url)
@@ -331,9 +336,13 @@ def qute_help(url):
 
     path = 'html/doc/{}'.format(urlpath)
     if not urlpath.endswith('.html'):
+        try:
+            bdata = utils.read_file(path, binary=True)
+        except OSError as e:
+            raise QuteSchemeOSError(e)
         mimetype, _encoding = mimetypes.guess_type(urlpath)
         assert mimetype is not None, url
-        return mimetype, utils.read_file(path, binary=True)
+        return mimetype, bdata
 
     try:
         data = utils.read_file(path)

--- a/qutebrowser/browser/qutescheme.py
+++ b/qutebrowser/browser/qutescheme.py
@@ -135,10 +135,10 @@ def data_for_url(url):
     Return:
         A (mimetype, data) tuple.
     """
-    # normalize path and strip redundant trailing slashes
-    norm_url = url.adjusted(QUrl.NormalizePathSegments | QUrl.StripTrailingSlash)
+    norm_url = url.adjusted(QUrl.NormalizePathSegments |
+            QUrl.StripTrailingSlash)
     if norm_url != url:
-       raise Redirect(norm_url)
+        raise Redirect(norm_url)
 
     path = url.path()
     host = url.host()

--- a/tests/end2end/features/navigate.feature
+++ b/tests/end2end/features/navigate.feature
@@ -18,6 +18,22 @@ Feature: Using :navigate
         And I run :navigate up with count 2
         Then data/navigate should be loaded
 
+    Scenario: Navigating up in qute://help/
+        When I open qute://help/commands.html
+        And I run :navigate up
+        Then qute://help/ should be loaded
+
+    Scenario: Navigating up by count in qute://help/
+        When I open qute://help/img/cheatsheet-big.png
+        And I run :navigate up with count 2
+        Then qute://help/ should be loaded
+
+    Scenario: Navigating up in qute://help/img/cheatsheet-big.png
+        When I open qute://help/img/cheatsheet-big.png
+        And I run :navigate up
+        Then "OSError while handling qute://* URL" should be logged
+        And "* url='qute://help/img'* LoadStatus.error" should be logged
+
     # prev/next
 
     Scenario: Navigating to previous page

--- a/tests/end2end/features/navigate.feature
+++ b/tests/end2end/features/navigate.feature
@@ -23,17 +23,6 @@ Feature: Using :navigate
         And I run :navigate up
         Then qute://help/ should be loaded
 
-    Scenario: Navigating up by count in qute://help/
-        When I open qute://help/img/cheatsheet-big.png
-        And I run :navigate up with count 2
-        Then qute://help/ should be loaded
-
-    Scenario: Navigating up in qute://help/img/cheatsheet-big.png
-        When I open qute://help/img/cheatsheet-big.png
-        And I run :navigate up
-        Then "OSError while handling qute://* URL" should be logged
-        And "* url='qute://help/img'* LoadStatus.error" should be logged
-
     # prev/next
 
     Scenario: Navigating to previous page

--- a/tests/end2end/features/qutescheme.feature
+++ b/tests/end2end/features/qutescheme.feature
@@ -63,6 +63,19 @@ Feature: Special qute:// pages
         And I hint with args "links normal" and follow a
         Then qute://help/quickstart.html should be loaded
 
+    Scenario: Opening a link with qute://help/index.html/..
+        When I open qute://help/index.html/.. without waiting
+        Then qute://help/ should be loaded
+
+    Scenario: Opening a link with qute://help/index.html/../
+        When I open qute://help/index.html/../ without waiting
+        Then qute://help/ should be loaded
+
+    Scenario: Opening a link with qute://help/img/cheatsheet-big.png/..
+        When I open qute://help/img/cheatsheet-big.png/.. without waiting
+        Then "OSError while handling qute://* URL" should be logged
+        And "* url='qute://help/img'* LoadStatus.error" should be logged
+
     # :history
 
     Scenario: :history without arguments

--- a/tests/end2end/features/qutescheme.feature
+++ b/tests/end2end/features/qutescheme.feature
@@ -64,15 +64,18 @@ Feature: Special qute:// pages
         Then qute://help/quickstart.html should be loaded
 
     Scenario: Opening a link with qute://help/index.html/..
-        When I open qute://help/index.html/.. without waiting
+        When the documentation is up to date
+        And I open qute://help/index.html/.. without waiting
         Then qute://help/ should be loaded
 
     Scenario: Opening a link with qute://help/index.html/../
-        When I open qute://help/index.html/../ without waiting
+        When the documentation is up to date
+        And I open qute://help/index.html/../ without waiting
         Then qute://help/ should be loaded
 
-    Scenario: Opening a link with qute://help/img/cheatsheet-big.png/..
-        When I open qute://help/img/cheatsheet-big.png/.. without waiting
+    Scenario: Opening a link with qute://help/img/
+        When the documentation is up to date
+        And I open qute://help/img/ without waiting
         Then "OSError while handling qute://* URL" should be logged
         And "* url='qute://help/img'* LoadStatus.error" should be logged
 

